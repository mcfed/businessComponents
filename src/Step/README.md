# Step 组件

## Usage

```jsx

    import {Step} from 'mcf-components';
    import FirstStep from "./firstStep";
    import SecondStep from "./secondStep";
    import ThirdStep from "./thirdStep";

    const steps = [
    {
        text: "第一步",
        description: "我是第一步",
        path: "first",
        component: FirstStep
    },
    {
        text: "第二步",
        description: "我是第二步",
        path: "second",
        component: SecondStep
    },
    {
        text: "第三步",
        description: "我是第三步",
        path: "third",
        component: ThirdStep
    }
    ];
    render(){
        return (
            <Step
                steps={steps}
                showCancel={true}
                cancelText="点击取消"
                // 传递router对象， 或者只传递localtion 和 push 方法
                {...props}
            />
        )
    }

    export withRouter(...)


    // firstep
    onSubmit = () => {
        if(true){
            // 跳转到第二步
            this.props.goToStep(2)
        }
    }

```

## Options

| 参数                    | 说明                       | 类型    | 默认值   |
| ----------------------- | -------------------------- | ------- | -------- |
| steps                   | 步骤组件数据               | array   | []       |
| showCancel              | 是否展示取消按钮           | boolean | true     |
| cancelText              | 取消按钮文字自定义         | string  | 取消     |
| finishText              | 最后一步完成按钮文字自定义 | string  | 完成     |
| showFinalLastStep       | 最后一步是否显示上一步按钮 | boolean | true     |
| finalSubmitFunctionName | 最后一步完成的响应方法名   | string  | onSubmit |
| backPath                | 取消按钮跳转的路由         | string  | /        |
