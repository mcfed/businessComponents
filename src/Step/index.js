import React, {Component} from 'react';
import {Button, Steps} from 'antd';

export const ref = React.createRef();

export default class Step extends Component {
  state = {
    currentIndex: 0,
    step: 1
  };

  componentWillMount() {
    const currentStepRoute = this.props.location.pathname.slice(1);

    let currentStepNum = 1;
    this.props.steps.map((v, index) => {
      /* istanbul ignore else */
      if (currentStepRoute.indexOf(v.path) > -1) {
        currentStepNum = index + 1;
      }
    });

    this.setState({
      currentIndex: currentStepNum - 1,
      step: currentStepNum
    });
  }

  goToStep = step => {
    const {
      props: {steps}
    } = this;
    this.props.history.push(steps[step - 1].path);
    this.setState({
      currentIndex: step - 1,
      step
    });
  };

  renderStepButtonGroups() {
    const {steps, showPrev, showCancel} = this.props;
    const {step, currentIndex} = this.state;
    return (
      <Button.Group>
        {step !== steps.length &&
        (typeof showCancel === 'boolean'
          ? showCancel
          : showCancel(currentIndex)) ? (
          <Button onClick={() => this.props.history.push(this.props.backPath)}>
            {this.props.cancelText}
          </Button>
        ) : (
          ''
        )}
        {(() => {
          switch (step) {
            case 1:
              return '';
            case steps.length:
              /* istanbul ignore next */
              return showPrev && this.props.showFinalLastStep ? (
                <Button onClick={() => this.goToStep(step - 1)}>
                  {'上一步'}
                </Button>
              ) : (
                ''
              );
            default:
              return showPrev ? (
                <Button onClick={() => this.goToStep(step - 1)}>
                  {'上一步'}
                </Button>
              ) : (
                ''
              );
          }
        })()}
        {step !== steps.length ? (
          <Button onClick={() => ref.current.onSubmit('handleSubmit')}>
            {'下一步'}
          </Button>
        ) : (
          ''
        )}
        {step === steps.length ? (
          <Button
            onClick={() =>
              /* istanbul ignore next */
              this.props.finalSubmitFunctionName
                ? ref.current &&
                  ref.current[this.props.finalSubmitFunctionName]()
                : ref.current.onSubmit('handleSubmit')
            }>
            {this.props.finishText}
          </Button>
        ) : (
          ''
        )}
      </Button.Group>
    );
  }

  render() {
    const {steps} = this.props;
    const {step, currentIndex} = this.state;
    /* istanbul ignore next */
    const renderDom = steps[currentIndex] || steps[0];
    return (
      <React.Fragment>
        <Steps current={currentIndex}>
          {steps.map(item => (
            <Steps.Step
              key={item.text}
              title={item.text}
              description={item.description}
            />
          ))}
        </Steps>

        <renderDom.component
          ref={ref}
          {...this.props}
          goToStep={this.goToStep}
          goToNext={() => this.goToStep(step + 1)}
        />

        {this.renderStepButtonGroups()}
      </React.Fragment>
    );
  }
}

Step.defaultProps = {
  steps: [],
  cancelText: '取消',
  finishText: '完成',
  showPrev: true,
  showFinalLastStep: true,
  showCancel: true,
  backPath: '/',
  location: {
    pathname: ''
  },
  history: {
    push() {}
  }
};
