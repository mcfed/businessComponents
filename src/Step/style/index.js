import 'antd/es/button/style';
import 'antd/es/steps/style';
