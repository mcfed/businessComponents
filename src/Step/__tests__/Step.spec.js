import {shallow} from 'enzyme';
import React from 'react';

import Step, {ref} from '../index';

class First extends React.Component {
  onSubmit() {
    this.props.goToStep(2);
  }
  render() {
    return 'i am First step';
  }
}

class Second extends React.Component {
  onSubmit() {
    this.props.goToStep(3);
  }
  render() {
    return 'i am Second step';
  }
}

class Third extends React.Component {
  onSubmit() {}
  render() {
    return 'i am Third step';
  }
}

const setup = props => {
  const wrapper = shallow(<Step {...props} />);

  return {
    wrapper,
    props
  };
};

const steps = [
  {
    text: '第一步',
    description: '我是第一步',
    path: 'first',
    component: First
  },
  {
    text: '第二步',
    description: '我是第二步',
    path: 'second',
    component: Second
  },
  {
    text: '第三步',
    description: '我是第三步',
    path: 'third',
    component: Third
  }
];

describe('快照测试', () => {
  it('全页快照', () => {
    const props = {
      steps: steps
    };
    const {wrapper} = setup(props);

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Step  render', () => {
  it('Step 正确渲染', () => {
    const {wrapper} = setup({steps: steps});
    expect(wrapper.find('First').exists()).toBe(true);
  });
});

describe('Step 点击事件', () => {
  delete window.location;
  window.location = {
    href: 'http://localhost/first',
    assign: url => {}
  };

  const props = {
    steps: steps,
    history: {
      push: jest.fn(path => {
        window.location.href = `http://localhost/${path}`;
      })
    }
  };
  it('点击下一步，能正确跳转，路由是否正确', () => {
    const {wrapper} = setup(props);
    expect(window.location.href.indexOf('first') > -1).toBe(true);
    ref.current = {
      onSubmit: () => {
        wrapper.instance().goToStep(2);
      }
    };
    wrapper
      .find('Button')
      .last()
      .props()
      .onClick();
    expect(wrapper.instance().props.history.push).toHaveBeenCalled();
    expect(window.location.href.indexOf('second') > -1).toBe(true);
    expect(wrapper.find('Button').length).toBe(3);
  });

  it('没有上一步按钮', () => {
    const {wrapper} = setup(Object.assign({}, {showPrev: false}, {...props}));
    expect(window.location.href.indexOf('second') > -1).toBe(true);
    ref.current = {
      onSubmit: () => {
        wrapper.instance().goToStep(2);
      }
    };
    expect(wrapper.find('Button').length).toBe(2);
  });

  it('点击上一步，能正确跳转，路由是否正确', () => {
    const {wrapper} = setup(props);
    expect(window.location.href.indexOf('second') > -1).toBe(true);
    ref.current = {
      onSubmit: () => {
        wrapper.instance().goToStep(1);
      }
    };
    wrapper
      .find('Button')
      .at(1)
      .props()
      .onClick();
    expect(wrapper.instance().props.history.push).toHaveBeenCalled();
    expect(window.location.href.indexOf('first') > -1).toBe(true);
  });
});

describe('Step 各参数是否生效', () => {
  it('steps 数据参数测试', () => {
    const {wrapper} = setup({
      steps: [
        {
          text: '第一步',
          description: '我是第一步',
          path: 'first',
          component: First
        },
        {
          text: '第二步',
          description: '我是第二步',
          path: 'second',
          component: Second
        }
      ]
    });
    expect(wrapper.find('Step').length).toBe(2);
    expect(
      wrapper
        .find('Step')
        .first()
        .props().title
    ).toBe('第一步');
    expect(
      wrapper
        .find('Step')
        .last()
        .props().description
    ).toBe('我是第二步');
    expect(wrapper.find('First').props().cancelText).toBe('取消');
    expect(wrapper.find('First').props().finishText).toBe('完成');
  });

  it('cancelText,finishText,showCancel 等其他参数测试', () => {
    const {wrapper} = setup({
      steps: [
        {
          text: '第一步',
          description: '我是第一步',
          path: 'first',
          component: First
        },
        {
          text: '第二步',
          description: '我是第二步',
          path: 'second',
          component: Second
        }
      ],
      cancelText: '我要取消',
      finishText: '返回',
      showCancel: false,
      backPath: 'index',
      showFinalLastStep: 'false',
      finalSubmitFunctionName: 'handleSubmit'
    });

    expect(wrapper.find('First').props().cancelText).toBe('我要取消');
    expect(wrapper.find('First').props().finishText).toBe('返回');
    expect(wrapper.find('First').props().backPath).toBe('index');
    expect(wrapper.find('First').props().showFinalLastStep).toBe('false');
    expect(wrapper.find('First').props().finalSubmitFunctionName).toBe(
      'handleSubmit'
    );
    expect(wrapper.find('Button').length).toBe(1);
    expect(
      wrapper
        .find('Button')
        .children()
        .text()
    ).toBe('下一步');
  });
});

describe('测试覆盖率', () => {
  it('最后一步,只有一步的情况', () => {
    const props = {
      steps: [
        {
          text: '第一步',
          description: '我是第一步',
          path: 'first',
          component: First
        }
      ],
      location: {
        pathname: ['', 'first']
      },
      showFinalLastStep: true
    };
    const {wrapper} = setup(props);
    ref.current = {
      onSubmit: () => {
        wrapper.instance().goToStep(1);
      }
    };
    wrapper
      .find('Button')
      .first()
      .props()
      .onClick();
  });

  it('最后一步，多步', () => {
    const props = {
      steps: [
        {
          text: '第一步',
          description: '我是第一步',
          path: 'first',
          component: First
        },
        {
          text: '第二步',
          description: '我是第二步',
          path: 'second',
          component: Second
        }
      ],
      location: {
        pathname: ['', 'first']
      },
      showFinalLastStep: true
    };
    const {wrapper} = setup(props);
    ref.current = {
      onSubmit: () => {
        wrapper.instance().goToStep(2);
      }
    };

    wrapper
      .find('Button')
      .first()
      .props()
      .onClick();
    wrapper
      .find('Button')
      .first()
      .props()
      .onClick();
  });

  it('最后一步，多步，有除了首尾两步之外的步数', () => {
    const props = {
      steps: [
        {
          text: '第一步',
          description: '我是第一步',
          path: 'first',
          component: First
        },
        {
          text: '第二步',
          description: '我是第二步',
          path: 'second',
          component: Second
        },
        {
          text: '第三步',
          description: '我是第三步',
          path: 'third',
          component: Third
        }
      ],
      location: {
        pathname: ['', 'first']
      },
      showFinalLastStep: true
    };
    const {wrapper} = setup(props);
    ref.current = {
      onSubmit: () => {
        wrapper.instance().goToStep(2);
      }
    };
    // console.log(wrapper.debug())
    // 点击第一步的下一步
    wrapper
      .find('Button')
      .first()
      .props()
      .onClick();
    // 点击第二步的上一步
    wrapper
      .find('Button')
      .first()
      .props()
      .onClick();
    // 点击第一步的取消
    wrapper
      .find('Button')
      .last()
      .props()
      .onClick();
  });
});
