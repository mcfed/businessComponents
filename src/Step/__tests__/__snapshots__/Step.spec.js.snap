// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 全页快照 1`] = `
<Fragment>
  <Steps
    current={0}
  >
    <Step
      description="我是第一步"
      key="第一步"
      title="第一步"
    />
    <Step
      description="我是第二步"
      key="第二步"
      title="第二步"
    />
    <Step
      description="我是第三步"
      key="第三步"
      title="第三步"
    />
  </Steps>
  <First
    backPath="/"
    cancelText="取消"
    finishText="完成"
    goToNext={[Function]}
    goToStep={[Function]}
    history={
      Object {
        "push": [Function],
      }
    }
    location={
      Object {
        "pathname": "",
      }
    }
    showCancel={true}
    showFinalLastStep={true}
    showPrev={true}
    steps={
      Array [
        Object {
          "component": [Function],
          "description": "我是第一步",
          "path": "first",
          "text": "第一步",
        },
        Object {
          "component": [Function],
          "description": "我是第二步",
          "path": "second",
          "text": "第二步",
        },
        Object {
          "component": [Function],
          "description": "我是第三步",
          "path": "third",
          "text": "第三步",
        },
      ]
    }
  />
  <ButtonGroup>
    <Button
      block={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
    >
      取消
    </Button>
    <Button
      block={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
    >
      下一步
    </Button>
  </ButtonGroup>
</Fragment>
`;
